### PR TITLE
ParseFilters get list of Outlink from parser

### PR DESCRIPTION
**Pull request by jnioche**
We might want the ParseFilters to have access to the Oulinks extracted by the parsers. This is useful when creating new outlinks programatically e.g. navigational links that can't be found in Javascript but also in order to attach arbitrary metadata for some outlinks. It can also be used to do some filtering from within the parsefilters.

The parserbolt currently filters and generates the metadata for the outlinks before calling the parsefillters. 

@DigitalPebble/committers-crawler   can you please review and comment?

### CODE REVIEW

**Comment by jakekdodd** [comment]
Haven't had a chance to test it, but everything looks fine to me!

Quick question...how does LinkContentHandler treat relative path links? It might be useful to put a helper method in Outlink that cals `URLUtil.resolveURL`.

File: core/src/main/java/com/digitalpebble/storm/crawler/bolt/ParserBolt.java [hunk 1]
```diff
@@ -264,9 +270,14 @@ public void execute(Tuple tuple) {
         }
 
         if (emitOutlinks) {
-            List<Link> links = linkHandler.getLinks();
-            if (!links.isEmpty()) {
-                emitOutlinks(tuple, url, metadata, links);
+            if (!outlinks.isEmpty()) {
```
Comment (GuiForget) [inline core/src/main/java/com/digitalpebble/storm/crawler/bolt/ParserBolt.java]: There isn't much value in that if check anymore

**Comment by jnioche** [comment]
thanks guys

> how does LinkContentHandler treat relative path links?

it does not resolve them and keeps them relative

> It might be useful to put a helper method in Outlink that cals URLUtil.resolveURL.

something like `static makeAbsolute(URL parentURL, String targetURL)`

why not. Calling URLUtil.resolveURL is probably just as easy.

File: core/src/main/java/com/digitalpebble/storm/crawler/bolt/ParserBolt.java (hunk 1 above)
Comment (jnioche) [inline core/src/main/java/com/digitalpebble/storm/crawler/bolt/ParserBolt.java]: fixed in [master 7d70198]

**Pull request merged**